// @shopify/shopify-api/rest/admin/2024-01/country.js
"use strict";
/***********************************************************************************************************************
* This file is auto-generated. If you have an issue, please create a GitHub issue.                                     *
***********************************************************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.Country = void 0;
const base_1 = require("../../base");
const types_1 = require("../../../lib/types");
const province_1 = require("./province");
class Country extends base_1.Base {
    static async find({ session, id, fields = null }) {
        const result = await this.baseFind({
            session: session,
            urlIds: { "id": id },
            params: { "fields": fields },
        });
        return result.data ? result.data[0] : null;
    }
    static async delete({ session, id }) {
        const response = await this.request({
            http_method: "delete",
            operation: "delete",
            session: session,
            urlIds: { "id": id },
            params: {},
        });
        return response ? response.body : null;
    }
    static async all({ session, since_id = null, fields = null, ...otherArgs }) {
        const response = await this.baseFind({
            session: session,
            urlIds: {},
            params: { "since_id": since_id, "fields": fields, ...otherArgs },
        });
        return response;
    }
    static async count({ session, ...otherArgs }) {
        const response = await this.request({
            http_method: "get",
            operation: "count",
            session: session,
            urlIds: {},
            params: { ...otherArgs },
            body: {},
            entity: null,
        });
        return response ? response.body : null;
    }
}
exports.Country = Country;
Country.apiVersion = types_1.ApiVersion.January24;
Country.hasOne = {};
Country.hasMany = {
    "provinces": province_1.Province
};
Country.paths = [
    { "http_method": "delete", "operation": "delete", "ids": ["id"], "path": "countries/<id>.json" },
    { "http_method": "get", "operation": "count", "ids": [], "path": "countries/count.json" },
    { "http_method": "get", "operation": "get", "ids": [], "path": "countries.json" },
    { "http_method": "get", "operation": "get", "ids": ["id"], "path": "countries/<id>.json" },
    { "http_method": "post", "operation": "post", "ids": [], "path": "countries.json" },
    { "http_method": "put", "operation": "put", "ids": ["id"], "path": "countries/<id>.json" }
];
Country.resourceNames = [
    {
        "singular": "country",
        "plural": "countries"
    }
];
