// @shopify/shopify-api/rest/admin/2024-01/shop.js
"use strict";
/***********************************************************************************************************************
* This file is auto-generated. If you have an issue, please create a GitHub issue.                                     *
***********************************************************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.Shop = void 0;
const base_1 = require("../../base");
const types_1 = require("../../../lib/types");
class Shop extends base_1.Base {
    static async current({ session, fields = null, ...otherArgs }) {
        const result = await this.baseFind({
            session: session,
            urlIds: {},
            params: { "fields": fields, ...otherArgs },
        });
        return result.data ? result.data[0] : null;
    }
    static async all({ session, fields = null, ...otherArgs }) {
        const response = await this.baseFind({
            session: session,
            urlIds: {},
            params: { "fields": fields, ...otherArgs },
        });
        return response;
    }
}
exports.Shop = Shop;
Shop.apiVersion = types_1.ApiVersion.January24;
Shop.hasOne = {};
Shop.hasMany = {};
Shop.paths = [
    { "http_method": "get", "operation": "get", "ids": [], "path": "shop.json" }
];
Shop.resourceNames = [
    {
        "singular": "shop",
        "plural": "shops"
    }
];
