// @shopify/admin-api-client/dist/graphql/client.js
'use strict';

var graphqlClient = require('../graphql-client/dist/graphql-client/graphql-client.js');
var validations = require('../graphql-client/dist/api-client-utilities/validations.js');
var apiVersions = require('../graphql-client/dist/api-client-utilities/api-versions.js');
var utilities = require('../graphql-client/dist/api-client-utilities/utilities.js');
var constants = require('../constants.js');
var validations$1 = require('../validations.js');

function createAdminApiClient({ storeDomain, apiVersion, accessToken, userAgentPrefix, retries = 0, customFetchApi, logger, }) {
    const currentSupportedApiVersions = apiVersions.getCurrentSupportedApiVersions();
    const storeUrl = validations.validateDomainAndGetStoreUrl({
        client: constants.CLIENT,
        storeDomain,
    });
    const baseApiVersionValidationParams = {
        client: constants.CLIENT,
        currentSupportedApiVersions,
        logger,
    };
    validations$1.validateServerSideUsage();
    validations.validateApiVersion({
        client: constants.CLIENT,
        currentSupportedApiVersions,
        apiVersion,
        logger,
    });
    validations$1.validateRequiredAccessToken(accessToken);
    const apiUrlFormatter = generateApiUrlFormatter(storeUrl, apiVersion, baseApiVersionValidationParams);
    const config = {
        storeDomain: storeUrl,
        apiVersion,
        accessToken,
        headers: {
            "Content-Type": constants.DEFAULT_CONTENT_TYPE,
            Accept: constants.DEFAULT_CONTENT_TYPE,
            [constants.ACCESS_TOKEN_HEADER]: accessToken,
            "User-Agent": `${userAgentPrefix ? `${userAgentPrefix} | ` : ""}${constants.CLIENT} v${constants.DEFAULT_CLIENT_VERSION}`,
        },
        apiUrl: apiUrlFormatter(),
        userAgentPrefix,
    };
    const graphqlClient$1 = graphqlClient.createGraphQLClient({
        headers: config.headers,
        url: config.apiUrl,
        retries,
        customFetchApi,
        logger,
    });
    const getHeaders = utilities.generateGetHeaders(config);
    const getApiUrl = generateGetApiUrl(config, apiUrlFormatter);
    const getGQLClientParams = utilities.generateGetGQLClientParams({
        getHeaders,
        getApiUrl,
    });
    const client = {
        config,
        getHeaders,
        getApiUrl,
        fetch: (...props) => {
            return graphqlClient$1.fetch(...getGQLClientParams(...props));
        },
        request: (...props) => {
            return graphqlClient$1.request(...getGQLClientParams(...props));
        },
    };
    return Object.freeze(client);
}
function generateApiUrlFormatter(storeUrl, defaultApiVersion, baseApiVersionValidationParams) {
    return (apiVersion) => {
        if (apiVersion) {
            validations.validateApiVersion({
                ...baseApiVersionValidationParams,
                apiVersion,
            });
        }
        const urlApiVersion = (apiVersion ?? defaultApiVersion).trim();
        return `${storeUrl}/admin/api/${urlApiVersion}/graphql.json`;
    };
}
function generateGetApiUrl(config, apiUrlFormatter) {
    return (propApiVersion) => {
        return propApiVersion ? apiUrlFormatter(propApiVersion) : config.apiUrl;
    };
}

exports.createAdminApiClient = createAdminApiClient;
